### PR TITLE
av-store: use determine_new_blocks

### LINKED ISSUE #3300: Availability Store should use determine_new_blocks on import notifications
These logs: "Candidate included without being backed?", can stem from missing block import notifications and `CandidateEvent::Backed`. Missing these and other candidate events will lead to storage leaks over time as well. 

#3261 implements a utility for backwards iteration, so it should be easy to plug this into the availability store now.

The only thing is that it relies on both a lower bound number and an `is_known` function which might need some changes in the av-store schema.

### PR DESCRIPTION
**Pull request by ordian**
Fixes #3300.

- [x] tests

### CODE REVIEW

File: node/core/av-store/src/lib.rs [hunk 1]
```diff
@@ -732,9 +794,10 @@ fn note_block_included(
 				State::Unfinalized(at, mut within) => {
 					if let Err(i) = within.binary_search(&be_block) {
 						within.insert(i, be_block);
+						State::Unfinalized(at, within)
+					} else {
+						return Ok(());
```
Comment (ordian) [inline node/core/av-store/src/lib.rs:799]: if it's already there we don't need to do a db transaction, right?

File: node/core/av-store/src/lib.rs [hunk 2]
```diff
@@ -580,82 +617,107 @@ async fn process_block_activated(
...
 
-	// We need to request the number of validators based on the parent state, as that is the number of validators
-	// used to create this block.
+	// We need to request the number of validators based on the parent state,
+	// as that is the number of validators used to create this block.
 	let n_validators = {
 		let (tx, rx) = oneshot::channel();
 		ctx.send_message(
-			RuntimeApiMessage::Request(block_header.parent_hash, RuntimeApiRequest::Validators(tx)).into()
+			RuntimeApiMessage::Request(header.parent_hash, RuntimeApiRequest::Validators(tx)).into()
```
Comment (ordian) [inline node/core/av-store/src/lib.rs:688]: this looks very inefficient
we could query a session index and session info instead

File: node/core/av-store/src/lib.rs [hunk 3]
```diff
@@ -478,6 +483,36 @@ impl AvailabilityStoreSubsystem {
...
+			finalized_number: None,
+		}
+	}
+}
+
+#[derive(Default, Debug)]
+struct KnownBlocks {
+	by_hash: HashSet<Hash>,
+	by_number: BTreeSet<(BlockNumber, Hash)>,
+}
```
Comment (ordian) [inline node/core/av-store/src/lib.rs:496]: this is kept only in memory, so it might be causing a warn only on startup, but I think it's fine

Comment (rphmeier) [reply node/core/av-store/src/lib.rs:496]: testing my understanding: we're keeping the total set of unfinalized blocks in memory to test whether they're known?

Comment (rphmeier) [reply node/core/av-store/src/lib.rs:496]: perhaps `KnownUnfinalizedBlocks` would communicate this more clearly?

File: node/core/av-store/src/lib.rs (hunk 2 above)
Comment (rphmeier) [reply node/core/av-store/src/lib.rs:688]: `RollingSessionWindow` would work for this, I think.

File: node/core/av-store/src/lib.rs (hunk 3 above)
Comment (ordian) [reply node/core/av-store/src/lib.rs:496]: yes, we do the same in approval-distribution
> perhaps KnownUnfinalizedBlocks would communicate this more clearly?

:+1:

**Review by rphmeier — APPROVED** [review]

File: node/core/av-store/src/lib.rs (hunk 2 above)
Comment (ordian) [reply node/core/av-store/src/lib.rs:688]: was thinking the same and/or doing #2987

**Pull request merged**